keps/sig-scheduling/3101-pod-to-node-assignment/README.md

```diff
@@ -0,0 +1,872 @@
+know that this has succeeded?
+-->
+
+-   Allow users to force 1:1 pod-to-node assignment for a given pod
+
+### Non-Goals
+
+-   Find the best fit node for a given pod
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Add a new pod-level API that allows users to specify how many "worker" pods are
+allowed on a node. The API is implemented by the scheduler as a filter plugin
+and kubelet as an admission predicate.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wants a Pod to be scheduled alone on a node to avoid them from contending
+on resources (e.g. claiming all CPU cores available on a node). In other words,
+resource intensive workloads are needed to be scheduled on separate nodes.
+
+#### Story 2
+
+As a user, I want to deploy a Spark or MPI job such that only a single worker is
+running on each pod excluding daemonsets.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Users might be able to circumvent 1:1 pod to node assignments by creating a
+daemonset with node affinity to a specific node.
+
+#### Possible Mitigations
+
+-   Prevent users from creating daemonsets using a
+    [ResourceQuota](https://kubernetes.io/docs/concepts/policy/resource-quotas/)
+    -   Object count quota (`count/daemonsets`) can be used for this purpose.
+-   Restrict daemonsets creation to admins/specific users via kubernetes
+    [RBAC](https://kubernetes.io/docs/concepts/policy/resource-quotas/)
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The idea is to create a scheduler filter plugin such that given a pod and a
+node, a node is filtered out (as in considered no fit for the pod) if any of the
+following is true (assume the API is the annotation k8s.io/pod-per-node=true):
+
+-   The incoming pod is not a daemonset pod (by checking OwnerReference) and
```

Comment:
```suggestion
-   The incoming pod is not a daemonset or static pod (by checking OwnerReference) and
```

while this doesn't apply to the scheduler since the scheduler shouldn't be scheduling static pods, it applies to Kubelet since it will be importing the same logic.